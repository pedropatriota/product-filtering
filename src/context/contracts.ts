import { TProductsProps } from "../data/contracts";

export type TFilters = {
  property: string;
  operator: string;
  values: string[] | string;
};

export type TSelect = {
  property: {
    label: string;
    value: number | string | null;
    type: string;
    values?: string[];
  };
  operator: { label: string; value: number | string | null };
  values: { label: string; value: number | string | null }[];
};

export interface IHomeContextProps {
  filters: TFilters;
  select: TSelect;
  operators: (type: "string" | "number" | "enumerated") => {
    label: string;
    value: string;
  }[];
  properties: {
    label: string;
    value: number;
    type: string;
    values?: string[];
  }[];
  products: TProductsProps;
  handleSelectProperty: (value: TSelect | any) => void;
  handleSelectOperator: (value: TSelect | any) => void;
  handleSelectValues: (value: TSelect | any) => void;
  valuesSelectFilter: (id: string | number | null) =>
    | {
        label: string;
        value: string | number | null;
      }[];
}
