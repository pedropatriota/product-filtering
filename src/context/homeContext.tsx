import {
  FC,
  PropsWithChildren,
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import { getUniqueOrderedValues, getMethodsByOperators } from "../utils";
import { operatorsByProperties } from "../constants";

import datastore from "../data/dataStore";

import type { TSelect, TFilters, IHomeContextProps } from "./contracts";
import type { TProductsProps } from "../data/contracts";

export const HomeContext = createContext<IHomeContextProps>({
  properties: [],
  operators: () => [],
  products: [],
  select: {
    property: { label: "", value: null, type: "", values: [] },
    operator: { label: "", value: null },
    values: [{ label: "", value: null }],
  },
  filters: { property: 0, operator: "", values: [""] || "" },
  handleSelectProperty: () => () => {},
  handleSelectOperator: () => () => {},
  handleSelectValues: () => () => {},
  valuesSelectFilter: () => [],
  applyFilters: () => {},
  clearFilters: () => {},
});

const HomeContextProvider: FC<PropsWithChildren> = ({ children }) => {
  /*
   * States:
   */
  const [filters, setFilters] = useState<TFilters>({
    property: 0,
    operator: "",
    values: [] && "",
  });

  const [select, setSelect] = useState<TSelect>({
    property: { label: "", value: null, type: "", values: [] },
    operator: { label: "", value: null },
    values: [],
  });

  const [filteredTable, setFilteredTable] = useState<TProductsProps>([]);

  /*
   * Functions from datastore:
   */
  const products = useMemo(() => {
    return datastore.getProducts();
  }, [filters]);

  const properties = useMemo(
    () =>
      datastore.getProperties().map(({ id, name, type }) => {
        return {
          label: name,
          value: id,
          type,
        };
      }),
    []
  );

  const operators = useCallback(
    (type: "string" | "number" | "enumerated") => {
      const operatorsAllowed = operatorsByProperties[type];
      return datastore
        .getOperators()
        .filter(({ id }) => {
          return operatorsAllowed?.includes(id);
        })
        .map(({ id, text }) => {
          return {
            label: text,
            value: id,
          };
        });
    },
    [select.property?.type]
  );

  /*
   * Function responsible for filtering the table:
   */
  const filteredProducts = () => {
    let filtered = datastore.getProducts();
    const operators = datastore.getOperators();

    const property = properties.find(
      ({ value }) => value === filters?.property
    );
    const operator = operators.find((op) => op.id === filters.operator);

    filtered = filtered.filter(({ property_values }) => {
      const propertyValue = property_values.find(({ property_id }) => {
        return property_id === property?.value;
      })?.value;

      return getMethodsByOperators({ filters, propertyValue, operator });
    });

    setFilteredTable(filtered);
  };

  const applyFilters = () => filteredProducts();
  const clearFilters = () => {
    setFilters({ property: 0, operator: "", values: [] });
    setSelect({
      property: { label: "", value: null, type: "" },
      operator: { label: "", value: null },
      values: [],
    });
  };

  const handleSelect = useCallback(
    (newValue: TSelect | any, propertyName: string) => {
      setSelect({ ...select, [propertyName]: newValue });

      setFilters((currentFilters) => {
        if (Array.isArray(newValue)) {
          return {
            ...currentFilters,
            [propertyName]: newValue?.map(
              ({ value }: { value: string }) => value
            ),
          };
        } else {
          return {
            ...currentFilters,
            [propertyName]: newValue?.value,
          };
        }
      });
    },
    [filters, select, setFilters, setSelect]
  );

  const valuesSelectFilter = useCallback(
    (
      id: number | string | null
    ): { label: string; value: string | number | null }[] => {
      const options = products.map(({ property_values }) => {
        property_values.filter(({ property_id }) => {
          property_id === id;
        });
        return {
          label: property_values[id as number]?.value,
          value: property_values[id as number]?.value,
        };
      });

      return getUniqueOrderedValues(options);
    },
    []
  );

  /*
   * Effects to handle:
   * - if user change the property, it clears the operator and property values
   * - if user only change the operator, the property values is cleared
   * - if there is no filter applied, the table renders the getProducts()
   */

  useEffect(() => {
    setFilters({ ...filters, operator: "", values: [] });
    setSelect({ ...select, operator: { label: "", value: null }, values: [] });
  }, [filters.property]);

  useEffect(() => {
    setFilters({ ...filters, values: [] || "" });
    setSelect({ ...select, values: [] || "" });
  }, [filters.operator]);

  useEffect(() => {
    if (!(filters.values as string[] | number[])?.length || !filters.values) {
      setFilteredTable(products);
    }
  }, [filters.values]);

  const value = {
    properties,
    operators,
    products: filteredTable?.length > 0 ? filteredTable : products,
    select,
    filters,
    valuesSelectFilter,
    handleSelectProperty: (value: TSelect | any) =>
      handleSelect(value, "property"),
    handleSelectOperator: (value: TSelect | any) =>
      handleSelect(value, "operator"),
    handleSelectValues: (value: TSelect | any) => handleSelect(value, "values"),
    applyFilters,
    clearFilters,
  };

  return <HomeContext.Provider value={value}>{children}</HomeContext.Provider>;
};

export default HomeContextProvider;
