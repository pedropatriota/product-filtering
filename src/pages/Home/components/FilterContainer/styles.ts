import styled from "styled-components";
import { Trash2 } from "lucide-react";

export const FilterContainer = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  gap: 1rem;
  align-items: center;
`;

export const Button = styled.button`
  height: 2.4rem;
  width: 7.625rem;
  border: none;
  background-color: #b9c5f0;
  color: #000;
  border-radius: 0.5rem;
  transition: background-color 0.3s ease-in-out;
  text-transform: uppercase;
  font-weight: 600;
  cursor: pointer;

  &:hover {
    background-color: #8294d3;
  }

  &:disabled {
    background-color: #ccc;
    cursor: not-allowed;
  }
`;

export const ClearIcon = styled(Trash2)`
  cursor: pointer;
  stroke: red;
  width: 2rem;
`;
