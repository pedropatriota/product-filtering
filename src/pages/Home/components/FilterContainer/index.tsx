import { useContext } from "react";
import { Filter } from "../../../../components";
import { HomeContext } from "../../../../context/homeContext";

import * as Styled from "./styles";

const FilterContainer = () => {
  const {
    properties,
    operators,
    select,
    handleSelectOperator,
    handleSelectProperty,
    handleSelectValues,
    valuesSelectFilter,
    applyFilters,
    clearFilters,
  } = useContext(HomeContext);

  return (
    <Styled.FilterContainer>
      <Filter
        filter={select}
        handleFilter={[
          handleSelectProperty,
          handleSelectOperator,
          handleSelectValues,
        ]}
        label={["property", "operator", "values"]}
        options={[
          properties,
          operators(
            select?.property?.type as "string" | "number" | "enumerated"
          ),
          valuesSelectFilter(select?.property?.value),
        ]}
      />
      <Styled.Button disabled={!select.values} onClick={applyFilters}>
        Apply
      </Styled.Button>
      <Styled.ClearIcon
        role="button"
        aria-label="clear"
        onClick={clearFilters}
      />
    </Styled.FilterContainer>
  );
};

export default FilterContainer;
