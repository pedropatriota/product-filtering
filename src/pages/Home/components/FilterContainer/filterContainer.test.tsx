import { FC, PropsWithChildren } from "react";
import { renderHook } from "@testing-library/react-hooks";
import {
  render,
  fireEvent,
  waitFor,
  findAllByText,
} from "@testing-library/react";
import { HomeContext } from "../../../../context/homeContext";
import datastore from "../../../../data/dataStore";
import FilterContainer from ".";

const operators = jest.fn().mockReturnValue((type: string) => {
  if (type === "string")
    return [
      { label: "Equals", value: "equals" },
      { label: "Has Any Of", value: "any" },
      { label: "Has None Of", value: "none" },
      { label: "Is Any Of", value: "any" },
      { label: "Contains", value: "contains" },
    ];

  return [];
});

describe("FilterContainer Component", () => {
  const props = {
    filters: { property: 0, operator: "", values: [] },
    products: datastore.products,
    properties: [
      { label: "Product Name", value: 0, type: "string" },
      { label: "color", value: 1, type: "string" },
      { label: "weight (oz)", value: 2, type: "number" },
      { label: "category", value: 3, type: "enumerated" },
      { label: "wireless", value: 4, type: "enumerated" },
    ],
    operators,
    select: {
      property: { label: "", value: null, type: "", values: [] },
      operator: { label: "", value: "" },
      values: [{ label: "", value: null }],
    },
    handleSelectOperator: jest.fn(),
    handleSelectProperty: jest.fn(),
    handleSelectValues: jest.fn(),
    valuesSelectFilter: jest.fn(),
    applyFilters: jest.fn(),
    clearFilters: jest.fn(),
  };

  const CustomProvider: FC<PropsWithChildren> = ({ children }) => (
    <HomeContext.Provider value={{ ...props }}>{children}</HomeContext.Provider>
  );
  it("should display correctly", async () => {
    const { getAllByLabelText } = render(
      <CustomProvider>
        <FilterContainer />
      </CustomProvider>
    );

    const propertySelect = getAllByLabelText(/select/i)[0] as HTMLInputElement;
    const operatorSelect = getAllByLabelText(/select/i)[1] as HTMLInputElement;
    const valuesSelect = getAllByLabelText(/select/i)[2] as HTMLInputElement;

    expect(propertySelect.value).toBe("");
    expect(operatorSelect.value).toBe("");
    expect(valuesSelect?.value).toBe("");

    expect(operatorSelect).toBeDisabled();
    expect(valuesSelect).toBeDisabled();
  });

  it("should be able to select the params to filter the products", async () => {
    const { getAllByLabelText, getByText, findByText, rerender } = render(
      <CustomProvider>
        <FilterContainer />
      </CustomProvider>
    );

    const propertySelect = getAllByLabelText(/select/i)[0] as HTMLInputElement;

    fireEvent.keyDown(propertySelect, {
      key: "ArrowDown",
      keyCode: 40,
      code: 40,
    });
    const selectedProperty = getByText("Product Name");
    fireEvent.click(selectedProperty);
    expect(props.handleSelectProperty).toHaveBeenCalledWith(
      { label: "Product Name", type: "string", value: 0 },
      { action: "select-option", name: undefined, option: undefined }
    );
  });

  it("should filter the products when pressing the button apply", () => {
    const { getByRole } = render(
      <HomeContext.Provider
        value={{
          ...props,
          select: {
            property: { label: "Product Name", value: 0, type: "string" },
            operator: { label: "Equals", value: "equals" },
            values: [{ label: "Keyboard", value: "keyboard" }],
          },
        }}
      >
        <FilterContainer />
      </HomeContext.Provider>
    );

    const buttonApply = getByRole("button", { name: "Apply" });

    fireEvent.click(buttonApply);

    expect(props.applyFilters).toBeCalled();
  });

  it("should clear the filters when pressing the button clear", () => {
    const { getAllByRole } = render(
      <HomeContext.Provider
        value={{
          ...props,
          select: {
            property: { label: "Product Name", value: 0, type: "string" },
            operator: { label: "Equals", value: "equals" },
            values: [{ label: "Keyboard", value: "keyboard" }],
          },
        }}
      >
        <FilterContainer />
      </HomeContext.Provider>
    );

    const buttonClear = getAllByRole("button", {
      name: /clear/i,
    });

    fireEvent.click(buttonClear[0]);

    expect(props.clearFilters).toBeCalled();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <CustomProvider>
        <FilterContainer />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
