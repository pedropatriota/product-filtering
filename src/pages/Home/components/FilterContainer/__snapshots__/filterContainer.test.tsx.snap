// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterContainer Component should match snapshot 1`] = `
<div>
  <div
    class="sc-dkjKgF gLwynD"
  >
    <div
      class="sc-grXZZQ frvTkF"
      data-testid="filter"
    >
      <div>
        <label
          for="react-select-2-input"
        >
          Select property
        </label>
        <div
          class="sc-bcPKhP bpHLtS css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-14-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control css-13cymwt-control"
          >
            <div
              class="react-select__value-container react-select__value-container--has-value css-1fdsijx-ValueContainer"
            >
              <div
                class="react-select__single-value css-1dimb5e-singleValue"
              />
              <div
                class="react-select__input-container css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Select property"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  id="react-select-14-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__clear-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
              <span
                class="react-select__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <label
          for="react-select-2-input"
        >
          Select operator
        </label>
        <div
          class="sc-bcPKhP bpHLtS react-select--is-disabled css-3iigni-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-15-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control react-select__control--is-disabled css-16xfy0z-control"
          >
            <div
              class="react-select__value-container css-1fdsijx-ValueContainer"
            >
              <div
                class="react-select__placeholder css-1jqq78o-placeholder"
                id="react-select-15-placeholder"
              >
                Select...
              </div>
              <div
                class="react-select__input-container css-1h01tm3-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-15-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Select operator"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  disabled=""
                  id="react-select-15-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-894a34-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <label
          for="react-select-2-input"
        >
          Select one or more values
        </label>
        <div
          class="sc-bcPKhP bpHLtS react-select--is-disabled css-3iigni-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-16-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control react-select__control--is-disabled css-16xfy0z-control"
          >
            <div
              class="react-select__value-container react-select__value-container--is-multi css-1fdsijx-ValueContainer"
            >
              <div
                class="react-select__placeholder css-1jqq78o-placeholder"
                id="react-select-16-placeholder"
              >
                Select...
              </div>
              <div
                class="react-select__input-container css-1h01tm3-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-16-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Select one or more values"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  disabled=""
                  id="react-select-16-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-894a34-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="sc-hLclGa kqDAoA"
    >
      Apply
    </button>
    <svg
      aria-label="clear"
      class="sc-eDnVMP kZsety"
      fill="none"
      height="24"
      role="button"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 6h18"
      />
      <path
        d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
      />
      <path
        d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
      />
      <line
        x1="10"
        x2="10"
        y1="11"
        y2="17"
      />
      <line
        x1="14"
        x2="14"
        y1="11"
        y2="17"
      />
    </svg>
  </div>
</div>
`;
